File: visa-dev-samplecode-master/vdp-node/libs/visaapiclient.js
var request = require('request');
var fs = require('fs');
var XPayUtil = require('./xpayutil.js')
var config = require('../config/configuration.json');
var randomstring = require('randomstring');

function logRequest(requestBody, path) {
	console.log("URL : " + config.visaUrl + path);
	if (requestBody !== null && requestBody !== '') {
		console.log("Request Body : " + JSON.stringify(JSON.parse(requestBody),null,4));
	}
}

function logResponseBody(response, body) {
	console.log("Response Code: " + response.statusCode);
	console.log("Headers:");
	for(var item in response.headers) {
		console.log(item + ": " + response.headers[item]);
	}
	console.log("Body: "+ JSON.stringify(JSON.parse(body),null,4));
}

function getBasicAuthHeader(userId, password) {
	return 'Basic ' + new Buffer(userId + ':' + password).toString('base64');
}

function VisaAPIClient() {
	
}

VisaAPIClient.prototype.doMutualAuthRequest = function(path, requestBody, methodType, headers, callback) {
	
	var userId = config.userId ;
	var password = config.password;
	var keyFile = config.key;
	var certificateFile = config.cert;
	logRequest(requestBody, path);
	
	if (methodType === 'POST' || methodType === 'PUT') {
		headers['Content-Type'] = 'application/json';
	}

	headers['Accept'] = 'application/json';
	headers['Authorization'] = getBasicAuthHeader(userId, password);
	headers['x-correlation-id'] = randomstring.generate({length:12, charset: 'alphanumeric'}) + '_SC'
	request({
		uri : config.visaUrl + path,
		key: fs.readFileSync(keyFile),
		method : methodType,
		cert: fs.readFileSync(certificateFile),
		headers: headers,
		body: requestBody,
		timeout: 30000
	}, function(error, response, body) {
		if (!error) {
			logResponseBody(response, body);
			callback(null, response.statusCode);
		} else {
			console.log(error);
			callback(error);
		}
	});
};

VisaAPIClient.prototype.doXPayRequest = function(baseUri, resourcePath, queryParams, requestBody, methodType, headers, callback) {
	logRequest(requestBody, baseUri + resourcePath + '?' + queryParams);
	
	var xPayUtil = new XPayUtil();
	
	if (methodType === 'POST' || methodType === 'PUT') {
		headers['Content-Type'] = 'application/json';
	}
	
	headers['Accept'] = 'application/json';
	headers['x-pay-token'] = xPayUtil.getXPayToken(resourcePath, queryParams, requestBody);
	headers['x-correlation-id'] = randomstring.generate({length:12, charset: 'alphanumeric'}) + '_SC'
	request({
		uri : config.visaUrl + baseUri + resourcePath + '?' + queryParams,
		method : methodType,
		headers: headers,
		body: requestBody
	}, function(error, response, body) {
		if (!error) {
			logResponseBody(response, body);
			callback(null, response.statusCode);
		} else {
			console.log(error);
			callback(error);
		}
	});
};

module.exports = VisaAPIClient;

File: vdp-ruby/lib/visa_api_client.rb
require 'rest-client'
require 'yaml'
require File.expand_path('../x_pay_utils', __FILE__)

class VisaAPIClient
  
  def initialize
    @config = YAML.load_file('configuration.yml')
    @x_pay_util = XPayUtils.new
  end
  
  def getCorrelationId()
    # Passing correlation id header (x-correlation-id) is optional while making API calls.
    correlation_id = (0...12).map { (48 + rand(10)).chr }.join
    return correlation_id
  end
  
  def logRequest(test_info, url, request_body)
    puts test_info
    puts "URL : #{url}"
    if (request_body != nil && request_body != '')
      puts "Request Body : "
      puts request_body
    end
  end
  
  def logResponse(response)
    puts "Response Status : #{response.code.to_s}"
    puts "Response Headers : " 
    for header,value in response.headers
       puts "#{header.to_s} : #{value.to_s}"
    end
    puts "Response Body : " + JSON.pretty_generate(JSON.parse(response.body))
  end
  
  def doMutualAuthRequest(path, test_info, method_type, request_body, headers={})
    url = "#{@config['visaUrl']}#{path}"
    user_id = @config['userId']
    password = @config['password']
    key_path = @config['key']
    cert_path = @config['cert']
    correlation_id = getCorrelationId()
    logRequest(test_info, url, request_body)
    if method_type == 'post' || method_type == 'put'
      headers['Content-type'] = 'application/json'
    end
    headers['accept'] = 'application/json'
    headers['x-correlation-id'] = "#{correlation_id}_SC" 
    begin
      response = RestClient::Request.execute(
      :method => method_type,
      :url => url,
      :headers => headers,
      :payload => request_body,
      :user => user_id,
      :password => password,
      :ssl_client_key => OpenSSL::PKey::RSA.new(File.read(key_path)),
      :ssl_client_cert =>  OpenSSL::X509::Certificate.new(File.read(cert_path))
      )
      logResponse(response)
      return response.code.to_s
    rescue RestClient::ExceptionWithResponse => e
      logResponse(e.response)
      return e.response.code.to_s
    end
  end
  
  def doXPayTokenRequest(base_uri, resource_path, query_string, test_info, method_type, request_body, headers={})
      api_key = @config['apiKey']
      shared_secret = @config['sharedSecret']
      url = "#{@config['visaUrl']}#{base_uri}#{resource_path}?#{query_string}"
      correlation_id = getCorrelationId()
      logRequest(test_info, url, request_body)
      xpay_token = @x_pay_util.get_xpay_token(resource_path, query_string, request_body)
      if method_type == 'post' || method_type == 'put'
          headers['Content-type'] = 'application/json'
      end
      headers['accept'] = 'application/json'
      headers['x-correlation-id'] = "#{correlation_id}_SC"
      headers['x-pay-token'] = xpay_token 
      begin
        response = RestClient::Request.execute(
        :method => method_type,
        :url => url,
        :headers => headers,
        :payload => request_body
        )
        logResponse(response)
        return response.code.to_s
      rescue RestClient::ExceptionWithResponse => e
        logResponse(e.response)
        return e.response.code.to_s
      end
    end

end
